### DISCUSSION TITLE
Test flight development from Python library

### DISCUSSION THREAD

**Discussion by danielcmdeveloper** [bitcraze/discussions#1612] [General]
Good morning,

I am trying to develop an example where the CrazyFlie takes off and hovers. I have LPS installed but when I take off the CrazyFlie, it does not stay in place but gradually drifts away.
Can anyone help me with the Python code for this program?.

Best regards.
Thank you very much.

**Comment by knmcguire** [comment]
Hi! 

I'm assuming that you have gone through the Loco positioning system [getting started tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-loco-positioning-system/) and then [the flying part ](https://www.bitcraze.io/documentation/tutorials/getting-started-with-flying-using-lps/), although you could use the flight commander in cfclient too.

Did you go through all the steps? Is the crazyflie able to estimate it's position when you hold it in the hand?

**Reply by danielcmdeveloper** [reply]
Good morning ,
Yes, we have done all the configuration you mentioned.
Reviewing it again, we have a question. 

We see that the Cartesian axis of the origin is not at the position of the anchor 0 but it is in the central part (Attached image).
In our case, the reference axis is at the position of the anchor 0, that is to say at the coordinates (0,0,0,0).
We do not know if this is correct or if it is something that does not inflate enough.
On the other hand we have seen that the configuration of the anchors is (attached image):

![imagenEjeCartesianoDeReferencia](https://github.com/user-attachments/assets/fad73f56-7de3-4292-8d69-9f8f459c4cd1)


In this case we see that they are negative, actually we don't know if this is for an example or if the disposition should be like this.
We have followed this configuration (Attached image):

![valoresDeLasAnclas](https://github.com/user-attachments/assets/5b02a216-a41c-499e-b19e-b57fe3d5fbae)


In this case it is observed that the axis of origin is the position (0,0,0) in the anchor 0.
Are we configuring the LPS anchors correctly?


![ejeEstandar](https://github.com/user-attachments/assets/38443e1e-2106-4c9e-b812-1123b8919eee)
![nuestraConfig](https://github.com/user-attachments/assets/e57e7d63-871a-4efb-a42c-c2481c4e725e)


Thank you very much in advance.
Best regards
Dani :-)

**Reply by knmcguire** [reply]
Hi! So indeed the anchor positions should be what you input in it. Is the problem that the anchors do not accept the new values, or?

Technically I would rather advise to have the 0.0.0 position on the middle bottom of the anchor box, so like how it has been implemented before.

The fact that the Crazyflie thinks it is higher than it is... if unfortunately just a limitation of the loco positioning system. If the anchors are close to the ground and there is something in the ground that is interfering with the signals then the estimate can be off.  The quality of the estimate usually improves more in the center of the 'box' but that means that you'll need to take off with an higher take off height to make sure that is above it's current estimate.

**Reply by danielcmdeveloper** [reply]
Good morning, 
First of all, I would like to thank you for the prompt response and the time you take to assist me. 
I understand that the origin of coordinates should be at the bottom centre, that is to say that it would be as follows (I attach an image with the point of origin of coordinates):
![OrigenDeCoordenadas](https://github.com/user-attachments/assets/30499282-a56c-419c-ba1c-2cd207615536)

We understand that the lower anchors (using the example of anchors 0 and 7) their coordinates with respect to the X axis would be X0 (-2.475 , 0,0) and X7 (2.475,0,0) if I am not mistaken. Thus we would place the origin at the lower centre.

On the other hand, as you say, the anchors must be separated from the ground (we have seen in the documentation that at least 15 cm), if this is not the case: would there be an error in the measurement?

We are also having a problem, and it is that when we take off with MotionCommander the CrazyFlie does not stay still in place, it starts to move to one side and leaves the flight field, we do not know if this is due to a motor that can make a lower thrust and makes it drift or that the code is not correct.

Thank you very much for your time. 
Best regards
I hope you have a pleasant weekend.
 Daniel Caballero

**Reply by knmcguire** [reply]
Oh my you are right! haha we actually never noticed that it could be interpreted in that way! I guess it would otherwise be in the way of the Bitcraze logo 😄 

The reasoning why we say that the anchor should not be placed directly on the ground is that the ground can interfer with the quality and the antenna, so it's just an advise to get a better quality.

For your issue with flight, are you sure that you are initializing/starting up the Crazyflie with it's heading facing the postiive x axis? The loco positioning system doesn't give any yaw information so it's needs to infer that from it's IMU alone.

**Comment by danielcmdeveloper** [comment]
Good morning,
Thank you very much for your answer and your time.
So I understand that the placement of the origin of coordinates is not something that is of vital importance, as long as the management of the coordinates and the drone are correct.

As for the issue of the anchors close to the ground, we believe that in this case one of the problems may be that, the anchors are less than 5 cm from the ground so the deviation, from what you tell us, may be due to that.

As for the problem with MotionCommander, we think we will open a new thread to explain the problem and just focus on that.

Thank you very much for everything.
Regards
Daniel Caballero.

**Reply by knmcguire** [reply]
Yes you have understood correctly. I guess where the 0,0,0 origin is is also out of personal preference, but I think for things to work out at least, it is the handiest to have at least the z-axis on 0 where the drone is on the actual ground
